src/hazelcore/OcamlExtraction.re

```diff
@@ -0,0 +1,374 @@
+      switch (snd(ocaml_d2)) {
+      //can't determine type
+      | Hole => (
+          fst(ocaml_d1) ++ "::" ++ fst(ocaml_d2),
+          List(snd(ocaml_d1)),
+        )
+      //d2 is already a list
+      | List(t) =>
+        if (t == snd(ocaml_d1) || t == Hole) {
+          (
+            "(" ++ fst(ocaml_d1) ++ "::" ++ fst(ocaml_d2) ++ ")",
+            snd(ocaml_d2),
+          );
+        } else {
+          failwith("Exp: Cons, type inconsistent");
+        }
+      | _ => failwith("Exp: Cons")
+      };
+    | Inj(t, side, de) =>
+      //ocaml don't have injection
+      let ocaml_exp = extract(~ctx, ~de);
+      switch (t) {
+      | Sum(l, r) =>
+        switch (side) {
+        | L => (fst(ocaml_exp), l)
+        | R => (fst(ocaml_exp), r)
+        }
+      | _ => failwith("Exp: Injection")
+      };
+    | Pair(d1, d2) =>
+      let ocaml_d1 = extract(~ctx, ~de=d1);
+      let ocaml_d2 = extract(~ctx, ~de=d2);
+      (
+        "(" ++ fst(ocaml_d1) ++ ", " ++ fst(ocaml_d2) ++ ")",
+        Prod([snd(ocaml_d1), snd(ocaml_d2)]),
+      );
+    | Triv => ("()", Hole) //We have no unit Htype...
+    | ConsistentCase(cases) =>
+      switch (cases) {
+      // the int seems useless
+      | Case(de, rules, _) =>
+        let ocaml_de = extract(~ctx, ~de);
+        let ocaml_rules = rules_extract(~ctx, ~rules, ~pat_t=snd(ocaml_de));
+        let str =
+          "((match ("
+          ++ fst(ocaml_de)
+          ++ ") with \n"
+          ++ fst(ocaml_rules)
+          ++ ") : "
+          ++ Typ.extract(~t=snd(ocaml_rules))
+          ++ ")";
+        (str, snd(ocaml_rules));
+      }
+    // inconsistbranches is a case have inconsistent types
+    | InconsistentBranches(_, _, _, _) =>
+      failwith("Exp: Ocaml don't allow inconsistent branches")
+    | Cast(_, _, _) => failwith("Exp: cast not allowed")
+    | FailedCast(_, _, _) => failwith("Exp: cast not allowed")
+    | InvalidOperation(_, _) => failwith("Exp: Invalid Operation")
+    }
+  and rules_extract =
+      (~ctx: Contexts.t, ~rules: list(DHExp.rule), ~pat_t: HTyp.t)
+      : extract_result =>
+    switch (rules) {
+    | [] => ("", Hole) //no rules, won't happen actually
+    | [h] => rule_extract(~ctx, ~rule=h, ~pat_t)
+    | [h, ...t] =>
+      let head = rule_extract(~ctx, ~rule=h, ~pat_t);
+      let tail = rules_extract(~ctx, ~rules=t, ~pat_t);
+      // if (snd(head) == snd(tail)) {
+      if (HTyp.consistent(snd(head), snd(tail))) {
+        (fst(head) ++ "\n" ++ fst(tail), snd(head));
+      } else {
+        failwith("Exp: Case rules with inconsistent results");
+      };
+    }
+  and rule_extract =
+      (~ctx: Contexts.t, ~rule: DHExp.rule, ~pat_t: HTyp.t): extract_result =>
+    switch (rule) {
+    | Rule(dp, de) =>
+      let ocaml_dp = Pat.extract(~dp);
+      //we seems don't allow constructors as pattern
+      //we should add the patterns into the context
+      let rec update_ctx = (dp: DHPat.t, ctx: Contexts.t): Contexts.t =>
```

Comment:
you should be able to call into Statics for this, don't need to write it yourself here...

Reply:
What do you mean by "call into Statics"? I'm a little bit confused on that, does it means I need to refer to something?

Reply:
you can use the Statics_Pat module to generate the context for a pattern, since that is what the pattern type synthesis function does.

Reply:
I see the syn in Statics_Pat requires type UHPat.t, but the patterns I receive should be already of type DHPat.t. I don't know how I can use the module.

Reply:
oh right, I'm sorry.

you should write an type_of function in DHPat that generates both the type and ctx for a pattern, and then call that here.

Reply:
I think the type_of function for DHPat is not realistic, because there's "Wild" which needs to judge by context. I'll first write a function takes an expected type for a pattern and generate the type and update the ctx instead, I'll first store it in Extraction file, and let's see how we can generalize it later.

Reply:
ah yeah it should take the type as input, not produce it as output.